File: apps/ios/App/ViewModels/TaxonomyViewModels.swift
import Foundation
import Observation
import Factory
import Data
import Domain

/// Manage categories — ported from apps/web/app/settings/categories/page.tsx.
///
/// The tree itself is `Domain.categoryTree`, vector-tested. What is here is the
/// live query, the draft state for the add row, and the three writes.
///
/// **Not ported: the Auto-categorize card.** It drives
/// `apps/web/src/categorize/` — a 905-line on-device merchant-matching engine
/// with its own seed taxonomy, normaliser and semantic matcher. That is its own
/// port, not a corner of this screen, and it is the same engine
/// `transactions/new` defers. Tracked in PARITY_AUDIT.
@Observable
@MainActor
final class CategoriesViewModel {
    @ObservationIgnored
    @Injected(\.ledgerRepository) private var ledgerRepository
    @ObservationIgnored
    @Injected(\.authRepository) private var authRepository

    var search = "" { didSet { rebuild() } }
    var newName = ""
    var newKind = "expense" { didSet { rebuild() } }
    var newParentId = ""

    private(set) var nodes: [CategoryTreeNode] = []
    /// The add row's parent options: top-level categories of the kind being
    /// added. Web recomputes this from `newKind`, so switching kind changes the
    /// list — and can leave `newParentId` pointing at a category of the other
    /// kind, which `rebuild` clears.
    private(set) var parentOptions: [TaxonomyCategory] = []

    private var expanded: Set<String> = []
    private var all: [TaxonomyCategory] = []
    private var task: Task<Void, Never>?

    func isExpanded(_ id: String) -> Bool { expanded.contains(id) }

    func toggle(_ id: String) {
        if expanded.contains(id) { expanded.remove(id) } else { expanded.insert(id) }
        rebuild()
    }

    func start() {
        guard task == nil else { return }
        task = Task { [weak self] in
            guard let self else { return }
            do {
                for try await rows in try self.ledgerRepository.watchCategories() {
                    // Sorted here rather than in the repository: every other
                    // reader of watchCategories() wants name order, and this
                    // screen is the only one that groups by kind first, which is
                    // what web's `ORDER BY kind, name` does.
                    self.all = rows
                        .map { TaxonomyCategory(id: $0.id, name: $0.name, kind: $0.kind, parentId: $0.parentId) }
                        .sorted { ($0.kind, $0.name) < ($1.kind, $1.name) }
                    self.rebuild()
                }
            } catch { print("Error watching categories: \(error)") }
        }
    }

    func cancel() {
        task?.cancel()
        task = nil
    }

    private func rebuild() {
        nodes = categoryTree(all, search: search, expanded: expanded)
        parentOptions = all.filter { $0.parentId == nil && $0.kind == newKind }
        if !newParentId.isEmpty && !parentOptions.contains(where: { $0.id == newParentId }) {
            newParentId = ""
        }
    }

    var canAdd: Bool { !newName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

    func add() {
        let name = newName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }
        let kind = newKind
        let parentId = newParentId.isEmpty ? nil : newParentId
        newName = ""
        newParentId = ""
        Task { [weak self] in
            guard let self, let userId = await self.resolveUserId() else { return }
            _ = try? await self.ledgerRepository.createCategory(
                userId: userId, name: name, kind: kind, parentId: parentId
            )
        }
    }

    func rename(_ id: String, to name: String) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        Task { [weak self] in
            try? await self?.ledgerRepository.renameCategory(id: id, name: trimmed)
        }
    }

    func delete(_ id: String) {
        Task { [weak self] in
            try? await self?.ledgerRepository.deleteCategory(id: id)
        }
    }

    /// Spelled out rather than `currentUserId ?? (try? await ensureUser())`:
    /// `??`'s right-hand side is `@autoclosure` and cannot hold an `await`.
    private func resolveUserId() async -> String? {
        if let existing = authRepository.currentUserId { return existing }
        return try? await authRepository.ensureUser()
    }
}

/// Manage labels — ported from apps/web/app/settings/labels/page.tsx.
@Observable
@MainActor
final class LabelsViewModel {
    @ObservationIgnored
    @Injected(\.ledgerRepository) private var ledgerRepository
    @ObservationIgnored
    @Injected(\.authRepository) private var authRepository

    /// Web's default swatch for a new label and its fallback for one saved
    /// without a colour. The value is web's; it is a literal there too.
    static let defaultColor = "#b06a4f"

    var search = "" { didSet { rebuild() } }
    var newName = ""
    var newColor = LabelsViewModel.defaultColor

    private(set) var labels: [LabelRow] = []
    private var all: [LabelRow] = []
    private var task: Task<Void, Never>?

    var canAdd: Bool { !newName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

    func start() {
        guard task == nil else { return }
        task = Task { [weak self] in
            guard let self else { return }
            do {
                for try await rows in try self.ledgerRepository.watchLabels() {
                    self.all = rows
                    self.rebuild()
                }
            } catch { print("Error watching labels: \(error)") }
        }
    }

    func cancel() {
        task?.cancel()
        task = nil
    }

    private func rebuild() {
        // Web tests `!search`, so only the empty string turns filtering off —
        // a single space really does filter. Copied, not corrected, so the two
        // platforms and the browser agree.
        guard !search.isEmpty else {
            labels = all
            return
        }
        let needle = search.lowercased()
        labels = all.filter { $0.name.lowercased().contains(needle) }
    }

    func add() {
        let name = newName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }
        let color = newColor
        newName = ""
        Task { [weak self] in
            guard let self, let userId = await self.resolveUserId() else { return }
            _ = try? await self.ledgerRepository.createLabel(userId: userId, name: name, color: color)
        }
    }

    func save(_ id: String, name: String, color: String) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        Task { [weak self] in
            try? await self?.ledgerRepository.updateLabel(id: id, name: trimmed, color: color)
        }
    }

    func delete(_ id: String) {
        Task { [weak self] in
            try? await self?.ledgerRepository.deleteLabel(id: id)
        }
    }

    private func resolveUserId() async -> String? {
        if let existing = authRepository.currentUserId { return existing }
        return try? await authRepository.ensureUser()
    }
}

File: apps/android/app/src/main/java/com/sanvya/app/ui/taxonomy/TaxonomyViewModels.kt
package com.sanvya.app.ui.taxonomy

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sanvya.app.data.auth.AuthRepository
import com.sanvya.app.data.repository.LabelRow
import com.sanvya.app.data.repository.LedgerRepository
import com.sanvya.app.domain.taxonomy.CategoryTreeNode
import com.sanvya.app.domain.taxonomy.TaxonomyCategory
import com.sanvya.app.domain.taxonomy.categoryTree
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

data class CategoriesState(
    val nodes: List<CategoryTreeNode> = emptyList(),
    /**
     * The add row's parent options: top-level categories of the kind being
     * added. Web recomputes this from the selected kind, so switching kind
     * changes the list -- and can leave a parent selected that belongs to the
     * other kind, which is why the screen clears it.
     */
    val parentOptions: List<TaxonomyCategory> = emptyList(),
)

/**
 * Manage categories -- ported from apps/web/app/settings/categories/page.tsx.
 *
 * The tree itself is `domain.taxonomy.categoryTree`, vector-tested. What is
 * here is the live query, the draft state for the add row, and the three writes.
 *
 * **Not ported: the Auto-categorize card.** It drives
 * `apps/web/src/categorize/` -- a 905-line on-device merchant-matching engine
 * with its own seed taxonomy, normaliser and semantic matcher. That is its own
 * port, not a corner of this screen, and it is the same engine
 * `transactions/new` defers. Tracked in PARITY_AUDIT.
 *
 * Mirrors apps/ios/App/ViewModels/TaxonomyViewModels.swift.
 */
class CategoriesViewModel : ViewModel(), KoinComponent {
    private val ledgerRepository: LedgerRepository by inject()
    private val authRepository: AuthRepository by inject()

    private val _search = MutableStateFlow("")
    val search: StateFlow<String> = _search.asStateFlow()

    private val _expanded = MutableStateFlow<Set<String>>(emptySet())

    private val _newName = MutableStateFlow("")
    val newName: StateFlow<String> = _newName.asStateFlow()

    private val _newKind = MutableStateFlow("expense")
    val newKind: StateFlow<String> = _newKind.asStateFlow()

    private val _newParentId = MutableStateFlow("")
    val newParentId: StateFlow<String> = _newParentId.asStateFlow()

    fun setSearch(v: String) { _search.value = v }
    fun setNewName(v: String) { _newName.value = v }
    fun setNewParentId(v: String) { _newParentId.value = v }

    fun setNewKind(v: String) {
        _newKind.value = v
        // A parent of the OTHER kind is no longer offered, so a stale selection
        // would silently file the new category under an invisible parent.
        _newParentId.value = ""
    }

    fun toggle(id: String) {
        _expanded.value = _expanded.value.let { if (id in it) it - id else it + id }
    }

    private val all: StateFlow<List<TaxonomyCategory>> = ledgerRepository.watchCategories()
        .map { rows ->
            // Sorted here rather than in the repository: every other reader of
            // watchCategories() wants name order, and this screen is the only
            // one that groups by kind first, which is what web's
            // `ORDER BY kind, name` does.
            rows.map { TaxonomyCategory(it.id, it.name, it.kind, it.parentId) }
                .sortedWith(compareBy({ it.kind }, { it.name }))
        }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val state: StateFlow<CategoriesState> = combine(
        all, _search, _expanded, _newKind,
    ) { categories, search, expanded, kind ->
        CategoriesState(
            nodes = categoryTree(categories, search, expanded),
            parentOptions = categories.filter { it.parentId == null && it.kind == kind },
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), CategoriesState())

    fun add() {
        val name = _newName.value.trim()
        if (name.isEmpty()) return
        val kind = _newKind.value
        val parentId = _newParentId.value.ifEmpty { null }
        _newName.value = ""
        _newParentId.value = ""
        viewModelScope.launch {
            val userId = authRepository.currentUserId.value ?: return@launch
            runCatching { ledgerRepository.createCategory(userId, name, kind, parentId) }
        }
    }

    fun rename(id: String, name: String) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return
        viewModelScope.launch { runCatching { ledgerRepository.renameCategory(id, trimmed) } }
    }

    fun delete(id: String) {
        viewModelScope.launch { runCatching { ledgerRepository.deleteCategory(id) } }
    }
}

/** Manage labels -- ported from apps/web/app/settings/labels/page.tsx. */
class LabelsViewModel : ViewModel(), KoinComponent {
    private val ledgerRepository: LedgerRepository by inject()
    private val authRepository: AuthRepository by inject()

    companion object {
        /**
         * Web's default swatch for a new label and its fallback for one saved
         * without a colour. The value is web's; it is a literal there too.
         */
        const val DEFAULT_COLOR = "#b06a4f"
    }

    private val _search = MutableStateFlow("")
    val search: StateFlow<String> = _search.asStateFlow()

    private val _newName = MutableStateFlow("")
    val newName: StateFlow<String> = _newName.asStateFlow()

    private val _newColor = MutableStateFlow(DEFAULT_COLOR)
    val newColor: StateFlow<String> = _newColor.asStateFlow()

    fun setSearch(v: String) { _search.value = v }
    fun setNewName(v: String) { _newName.value = v }
    fun setNewColor(v: String) { _newColor.value = v }

    val labels: StateFlow<List<LabelRow>> = combine(
        ledgerRepository.watchLabels(), _search,
    ) { rows, search ->
        // Web tests `!search`, so only the empty string turns filtering off --
        // a single space really does filter. Copied, not corrected, so the two
        // platforms and the browser agree.
        if (search.isEmpty()) rows else rows.filter { it.name.lowercase().contains(search.lowercase()) }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun add() {
        val name = _newName.value.trim()
        if (name.isEmpty()) return
        val color = _newColor.value
        _newName.value = ""
        viewModelScope.launch {
            val userId = authRepository.currentUserId.value ?: return@launch
            runCatching { ledgerRepository.createLabel(userId, name, color) }
        }
    }

    fun save(id: String, name: String, color: String) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return
        viewModelScope.launch { runCatching { ledgerRepository.updateLabel(id, trimmed, color) } }
    }

    fun delete(id: String) {
        viewModelScope.launch { runCatching { ledgerRepository.deleteLabel(id) } }
    }
}
